### ISSUE TITLE
File never gets uploaded and stays with 0 size:  error in io.Copy: context canceled

### ISSUE THREAD

**Issue by movergan** [GoogleCloudPlatform/gcsfuse#1257]
**Describe the issue**
I have an FTP server for storing backups. The directory of the FTP server is mounted GoogleBucket. When a relatively large file is uploaded to FTP, it is not getting uploaded to the Bucket. 

I see the file on the server:
```
ll /var/www/vhosts/backup.ftp/backup/server/ | grep backup_2308031324.tar
-rw-r--r--. 1 ftpbackup psacln  5505863931 Aug  3 13:28 backup_2308031324.tar
```

And the same file on the bucket:
```
backup_2308031324.tar | 0 B | application/x-tar | 3 Aug 2023, 15:25:39 | Standard | 3 Aug 2023, 15:25:39 | Not public
```

I know that the file is not uploaded to FTP at once, part of the file is uploaded straight away and another part within a few seconds, so the total upload to FTP time could be 1-2 minutes. 

**To Collect more Debug logs**
1. Set ProFTPD server pointing to vhost where the bucket is mounted
2. Upload large file 

Debug logs:
```
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"gcs: Req            0x114: -\u003e CreateObject(\"server/backup_2308031324.tar\") (4.988118179s): error in io.Copy: context canceled\n","timestampSeconds":1691069313,"timestampNanos":378554214}
{"name":"root","levelname":"ERROR","severity":"ERROR","message":"FlushFile: interrupted system call, FileInode.Sync: SyncObject: Create: CreateObject: error in io.Copy: context canceled\n","timestampSeconds":1691069313,"timestampNanos":378626576}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00007bbc        connection.go:500] -\u003e Error: \"interrupted system call\"\n","timestampSeconds":1691069313,"timestampNanos":378682598}
{"name":"root","levelname":"ERROR","severity":"ERROR","message":"fuse: *fuseops.FlushFileOp error: interrupted system call\n","timestampSeconds":1691069313,"timestampNanos":378692147}
```

CLI: 
```
/usr/bin/gcsfuse --foreground --gid 1006 --debug_gcs --max-retry-duration 30m -o allow_other --uid 10001 --file-mode 644 --debug_fuse --max-retry-sleep 30m --log-file /var/log/gcsfuse.log --debug_fs --debug_http -o rw --dir-mode 755 backup-site-bucket /var/www/vhosts/backup.ftp/backup

Calling gcsfuse with arguments: --gid 1006 --debug_gcs --max-retry-duration 30m -o allow_other --uid 10001 --file-mode 644 --debug_fuse --max-retry-sleep 30m --log-file /var/log/gcsfuse.log --debug_fs --debug_http -o rw --dir-mode 755 backup-site-bucket /var/www/vhosts/backup.ftp/backup
```

**System (please complete the following information):**
 - OS: Rocky Linux release 8.8 (Green Obsidian)
 - Platform: GCE VM
 - Version: gcsfuse version 1.0.1 (Go version go1.20.5)

**Additional context**
My assumption is that context was canceled because the fuse is trying to upload the file to GoobleBucket and the file has not been fully uploaded to FTP.

The question is, how do I make sure that once the file is uploaded to FTP, it eventually ends up being copied to the bucket? I've set timeout to 30m but it seems to be not sufficient. 

**SLO:**
24 hrs to respond and 7 days to close the issue.

**Comment by raj-prince** [comment]
Hi @movergan, sorry for the delayed response!

Based on the fuse log `Error: \"interrupted system call\"\n"`, this is clear gcsfuse's FlushFile call is cancelled due to interrupt via the FTP server.

It's interesting that this is happening only for relatively large file. I tried to debug this by creating our own ftp server using `proftpd`, but not able to access the gcsfuse mounted directory on ftp server via ftp client.

I need more informations:
1. The complete gcsfuse-logs file - want to check the whole content is staged to ftp server before the final sync/flush call of gcsfuse.
2. If you have created your own FTP server, could you please share the configuration of the ftp server to reproduce the issue. Currently, when I try to access gcsfuse mounted directory, I get the below response:
```
ftp> ls gcs
227 Entering Passive Mode (10,128,0,79,171,249).
150 Opening ASCII mode data connection for file list
226 Transfer complete
```
I referred [this](https://cloudinfrastructureservices.co.uk/install-ftp-server-on-centos-8/) doc to create the ftp server. And used other machine as ftp client.
3. Till what file-size, upload works successfully.

Ideally, this interrupt should not come while uploading the content to gcsfuse.

Regards,
Prince Kumar.

**Comment by movergan** [comment]
Hello,

Thank you for getting back to me. The FTP I am using is provided by the Plesk, but I am not sure if there is anything specific about FTP itself. The nature of the backup is that it creates part of the backup, and uploaded it to FTP. Then it makes another part of the backup and continues uploading to the same file. Here is the log of the backup:
```
[2023-08-03 14:23:27.639|988674] INFO: Ftp init url ftps://127.0.0.1/backup/server/backup_2308031419.tar
[2023-08-03 14:23:27.639|988674] INFO: Open file ftp://127.0.0.1/backup/server/backup_2308031419.tar
[2023-08-03 14:23:27.639|988674] INFO: Curl output: * Hostname 127.0.0.1 was found in DNS cache
[2023-08-03 14:23:27.639|988674] INFO: Curl output: *   Trying 127.0.0.1:21...
[2023-08-03 14:23:27.639|988674] INFO: Curl output: * Connected to 127.0.0.1 (127.0.0.1) port 21 (#1)
[2023-08-03 14:23:27.680|988674] INFO: Curl output: < 220 ProFTPD Server (ProFTPD) [127.0.0.1]
[2023-08-03 14:23:27.680|988674] INFO: Curl output: > AUTH SSL
[2023-08-03 14:23:27.680|988674] INFO: Curl output: < 234 AUTH SSL successful
[2023-08-03 14:23:27.681|988674] INFO: Curl output: * SSL re-using session ID
[2023-08-03 14:23:27.681|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS handshake, Client hello (1):
[2023-08-03 14:23:27.683|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Server hello (2):
[2023-08-03 14:23:27.683|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
[2023-08-03 14:23:27.683|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Certificate (11):
[2023-08-03 14:23:27.683|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, CERT verify (15):
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Finished (20):
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS handshake, Finished (20):
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * Server certificate:
[2023-08-03 14:23:27.684|988674] INFO: Curl output: *  subject: CN=stage.example.com
[2023-08-03 14:23:27.684|988674] INFO: Curl output: *  start date: Jul 15 14:29:08 2023 GMT
[2023-08-03 14:23:27.684|988674] INFO: Curl output: *  expire date: Oct 13 14:29:07 2023 GMT
[2023-08-03 14:23:27.684|988674] INFO: Curl output: *  issuer: C=US; O=Let's Encrypt; CN=R3
[2023-08-03 14:23:27.684|988674] INFO: Curl output: *  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
[2023-08-03 14:23:27.684|988674] INFO: Curl output: > USER username
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
[2023-08-03 14:23:27.684|988674] INFO: Curl output: * old SSL session ID is stale, removing
[2023-08-03 14:23:27.685|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
[2023-08-03 14:23:27.685|988674] INFO: Curl output: * old SSL session ID is stale, removing
[2023-08-03 14:23:27.685|988674] INFO: Curl output: < 331 Password required for username
[2023-08-03 14:23:27.685|988674] INFO: Curl output: > PASS *****
[2023-08-03 14:23:27.727|988674] INFO: Curl output: < 230 User username logged in
[2023-08-03 14:23:27.728|988674] INFO: Curl output: > PBSZ 0
[2023-08-03 14:23:27.728|988674] INFO: Curl output: < 200 PBSZ 0 successful
[2023-08-03 14:23:27.728|988674] INFO: Curl output: > PROT P
[2023-08-03 14:23:27.728|988674] INFO: Curl output: < 200 Protection set to Private
[2023-08-03 14:23:27.728|988674] INFO: Curl output: > PWD
[2023-08-03 14:23:27.728|988674] INFO: Curl output: < 257 "/" is the current directory
[2023-08-03 14:23:27.728|988674] INFO: Curl output: * Entry path is '/'
[2023-08-03 14:23:27.728|988674] INFO: Curl output: > OPTS UTF8 ON
[2023-08-03 14:23:27.728|988674] INFO: Curl output: * ftp_perform ends with SECONDARY: 0
[2023-08-03 14:23:27.729|988674] INFO: Curl output: < 200 UTF8 set to on
[2023-08-03 14:23:27.729|988674] INFO: Curl output: > CWD backup
[2023-08-03 14:23:27.730|988674] INFO: Curl output: < 250 CWD command successful
[2023-08-03 14:23:27.730|988674] INFO: Curl output: > CWD server
[2023-08-03 14:23:27.758|988674] INFO: Curl output: < 250 CWD command successful
[2023-08-03 14:23:27.758|988674] INFO: Curl output: > EPSV
[2023-08-03 14:23:27.758|988674] INFO: Curl output: * Connect data stream passively
[2023-08-03 14:23:27.759|988674] INFO: Curl output: < 229 Entering Extended Passive Mode (|||58659|)
[2023-08-03 14:23:27.759|988674] INFO: Curl output: * Connecting to 127.0.0.1 (127.0.0.1) port 58659
[2023-08-03 14:23:27.759|988674] INFO: Curl output: *   Trying 127.0.0.1:58659...
[2023-08-03 14:23:27.760|988674] INFO: Curl output: * Connected to 127.0.0.1 (127.0.0.1) port 21 (#1)
[2023-08-03 14:23:27.760|988674] INFO: Curl output: * SSL re-using session ID
[2023-08-03 14:23:27.760|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS handshake, Client hello (1):
[2023-08-03 14:23:27.760|988674] INFO: Curl output: > TYPE I
[2023-08-03 14:23:27.839|988674] INFO: Curl output: < 200 Type set to I
[2023-08-03 14:23:27.839|988674] INFO: Curl output: > SIZE backup_2308031419.tar
[2023-08-03 14:23:32.977|988674] INFO: Curl output: < 213 5505941095
[2023-08-03 14:23:32.977|988674] INFO: Curl output: * Instructs server to resume from offset 5502669824
[2023-08-03 14:23:32.977|988674] INFO: Curl output: > REST 5502669824
[2023-08-03 14:23:33.982|988674] INFO: Curl output: < 350 Restarting at 5502669824. Send STORE or RETRIEVE to initiate transfer
[2023-08-03 14:23:33.982|988674] INFO: Curl output: > RETR backup_2308031419.tar
[2023-08-03 14:23:33.986|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Server hello (2):
[2023-08-03 14:23:33.987|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
[2023-08-03 14:23:33.987|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Finished (20):
[2023-08-03 14:23:33.987|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
[2023-08-03 14:23:33.987|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS handshake, Finished (20):
[2023-08-03 14:23:33.987|988674] INFO: Curl output: * SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
[2023-08-03 14:23:33.987|988674] INFO: Curl output: * Server certificate:
[2023-08-03 14:23:33.987|988674] INFO: Curl output: *  subject: CN=stage.altenar.com
[2023-08-03 14:23:33.987|988674] INFO: Curl output: *  start date: Jul 15 14:29:08 2023 GMT
[2023-08-03 14:23:33.987|988674] INFO: Curl output: *  expire date: Oct 13 14:29:07 2023 GMT
[2023-08-03 14:23:33.988|988674] INFO: Curl output: *  issuer: C=US; O=Let's Encrypt; CN=R3
[2023-08-03 14:23:33.988|988674] INFO: Curl output: *  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
[2023-08-03 14:23:33.988|988674] INFO: Curl output: < 150 Opening BINARY mode data connection for backup_2308031419.tar (3271271 bytes)
[2023-08-03 14:23:33.988|988674] INFO: Curl output: * Maxdownload = -1
[2023-08-03 14:23:33.988|988674] INFO: Curl output: * Getting file with size: 3271271
[2023-08-03 14:23:33.988|988674] INFO: Curl output: * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
[2023-08-03 14:23:33.988|988674] INFO: Curl output: * old SSL session ID is stale, removing
[2023-08-03 14:23:33.990|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS alert, close notify (256):
[2023-08-03 14:23:33.990|988674] INFO: Curl output: * Closing connection 1
[2023-08-03 14:23:33.990|988674] INFO: Curl output: * TLSv1.3 (OUT), TLS alert, close notify (256):
```

I've also attached a full log gcsfuse-logs:
[gcsfuse.log](https://github.com/GoogleCloudPlatform/gcsfuse/files/12283129/gcsfuse.log)

Why backup size matters because the first part is being uploaded to the bucket when the second part from FTP arrives. If a file was too small, upload time would've been instant and it would've been difficult to reproduce this behavior. 

To sum up:

1. First portion of a large file uploaded to FTP
2. File transfer to bucket begins
3. Second part of the large file arrives
4. Transfer canceled. 

My question would be, why when the FTP upload is finished, the file is not getting uploaded to the bucket ever? If I rename this file on the disk, it appears in the bucket within a few minutes.

**Comment by raj-prince** [comment]
Hey @movergan, thanks for the details!

To answer your question, GCSFuse uploads the file as object in GCS bucket in two steps:

1. It creates a local file to write content locally using `WriteFile` method.
2. And then when GCSFuse gets FlushFile signal from kernel, it syncs the file content as object in GCS bucket.

In your case, if we analyse a smaller part of logs (collected from :
1. `Op 0x00005ee6 - WriteFile (inode 9, PID 0, handle 71, offset 5505761280, 179815 bytes)\n","timestampSeconds":1691072599,` - the last written chunk on the local file.
2. `Op 0x00005eea - FlushFile (inode 9, PID 986341)\n","timestampSeconds":1691072599` - Flush file signal to upload it to bucket. This came just after the `WriteFile` call. Check - `timestampSeconds`
3. `Op 0x00005eeb - interrupt (fuseid 0x00005eea)\n","timestampSeconds":1691072604` - After 5 seconds, interrupt signal for the FlushFile go-routine came. See id - `0x00005eea`.
4. `FlushFile: interrupted system call, FileInode.Sync: SyncObject: Create: CreateObject: error in io.Copy: context canceled\n","timestampSeconds":1691072604` - This happened due interrupt in the 4th step.
5. `Op 0x00005eea - Error: \"interrupted system call\"\n","timestampSeconds":1691072604` - This is response to the kernel back for the FlushFile call.

So, we need to debug 3rd step - why it came just after 5 seconds after the FlushFile call?

- I don't see any error in gcsfuse, previous to FlushFile call, which is one of the reasons of the interrupt.
- As this is coming from the app, I am not able to understand why it's coming.


**Analysed chunk of the logs:**
```JSON
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005ee6        connection.go:416] \u003c- WriteFile (inode 9, PID 0, handle 71, offset 5505761280, 179815 bytes)\n","timestampSeconds":1691072599,"timestampNanos":933696893}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005ee8        connection.go:416] \u003c- SetInodeAttributes (inode 9, PID 986341, mtime 2023-08-03 14:23:19.826383814 +0000 UTC)\n","timestampSeconds":1691072599,"timestampNanos":933800683}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005ee6        connection.go:498] -\u003e OK ()\n","timestampSeconds":1691072599,"timestampNanos":934013696}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005ee8        connection.go:498] -\u003e OK ()\n","timestampSeconds":1691072599,"timestampNanos":934043003}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005eea        connection.go:416] \u003c- FlushFile (inode 9, PID 986341)\n","timestampSeconds":1691072599,"timestampNanos":934150767}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"gcs: Req             0x47: \u003c- StatObject(\"server/backup_2308031419.tar\")\n","timestampSeconds":1691072599,"timestampNanos":939205137}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"gcs: Req             0x47: -\u003e StatObject(\"server/backup_2308031419.tar\") (30.40699ms): OK\n","timestampSeconds":1691072599,"timestampNanos":968282605}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"gcs: Req             0x48: \u003c- CreateObject(\"server/backup_2308031419.tar\")\n","timestampSeconds":1691072599,"timestampNanos":968370625}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005eeb        connection.go:416] \u003c- interrupt (fuseid 0x00005eea)\n","timestampSeconds":1691072604,"timestampNanos":828952385}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"gcs: Req             0x48: -\u003e CreateObject(\"server/backup_2308031419.tar\") (4.862939971s): error in io.Copy: context canceled\n","timestampSeconds":1691072604,"timestampNanos":831321808}
{"name":"root","levelname":"ERROR","severity":"ERROR","message":"FlushFile: interrupted system call, FileInode.Sync: SyncObject: Create: CreateObject: error in io.Copy: context canceled\n","timestampSeconds":1691072604,"timestampNanos":831408701}
{"name":"root","levelname":"DEBUG","severity":"DEBUG","message":"fuse_debug: Op 0x00005eea        connection.go:500] -\u003e Error: \"interrupted system call\"\n","timestampSeconds":1691072604,"timestampNanos":831494023}
{"name":"root","levelname":"ERROR","severity":"ERROR","message":"fuse: *fuseops.FlushFileOp error: interrupted system call\n","timestampSeconds":1691072604,"timestampNanos":831512578}
```

Thanks!

**Comment by movergan** [comment]
It is worth mentioning that it always fails in 5 sec. Please let me know if there is anything I can help you with.

**Comment by Tulsishah** [comment]
Hi @movergan ,

Sorry for the late response.
I tried to reproduce the issue, but I am having permission issue on accessing mounted directory. 

There are few analysis which I did,

1. The FTP server may be overloaded. If the FTP server is under a lot of load, it may not be able to handle the request to write a large file. This can happen if the server is being used by a lot of users or if it is trying to process a lot of requests at the same time.
-   Reduce the load on the FTP server. If the FTP server is overloaded, you can try reducing the load by reducing the number of users who are using it or by reducing the number of requests that are being processed at the same time.

2. The FTP client may be using too many connections. The FTP client may be trying to use too many connections to the FTP server. This can happen if the client is trying to upload a large file in multiple parts or if it is trying to upload multiple files at the same time.
- You can increase --max-conn-per-host to some larger value for gcsfuse.
- The default maximum number of connections for an FTP server is usually 10.(You can try by increasing limit)

3. The network connection between the FTP client and the FTP server may be unstable. If the network connection is unstable, it can cause the transfer of the file to be interrupted. This can happen if the network is experiencing congestion or if there is a problem with the router or switch.
- Check the network connection. If the network connection is unstable, you can try checking the network connection and resolving any problems that you find.


Thanks,
Tulsi Shah

**Comment by raj-prince** [comment]
@movergan 

Based on your attempt to increase the timeout to 30 seconds, I infer that you are referring to the FTP client-to-server communication.
I would like to know if there is a timeout for an operation between the FTP server trying to write the file on vFS (mounted via gcsfuse). If so, can we try increasing it?

-Prince

**Comment by raj-prince** [comment]
Hi @movergan,

That is the expected behavior. When GCSFuse cancels an operation due to an interrupt, the application must handle the interrupt error returned for the canceled operation.

We have also discussed this issue as a team and have decided not to support this workflow (mounting GCSFuse on an FTP server) at this time.

We apologize for any inconvenience this may cause. Closing this issue now.


Thanks,
Prince.
